borg/helpers.py

```diff
@@ -151,7 +152,13 @@ def load(cls, repository, key=None, force_tam_not_required=False):
     def write(self):
         if self.key.tam_required:
             self.config[b'tam_required'] = True
-        self.timestamp = datetime.utcnow().isoformat()
+        # self.timestamp needs to be strictly monotonically increasing. Clocks often are not set correctly
+        if self.timestamp is None:
+            self.timestamp = datetime.utcnow().isoformat()
+        else:
+            prev_ts = datetime.strptime(self.timestamp, "%Y-%m-%dT%H:%M:%S.%f")
+            incremented = (prev_ts + timedelta(microseconds=1)).isoformat()
+            self.timestamp = max(incremented, datetime.utcnow().isoformat())
```

Comment:
Sorry for jumping all around.
Rollback prevention is via
Cache.__init__ -> self.security_manager.assert_secure -> SecurityManager.assert_no_manifest_replay
After that has executed the next line in __init__ is the cache sync detection. It uses self.manifest.id != self.manifest_id which now only has to ensure that borg always rebuilds the cache if anything could possibly changed the repository state (including check, recreate, etc)